File: frontend/src/services/websocket.ts
import { io, Socket } from 'socket.io-client';
import { WS_URL } from '../utils/constants';

class WebSocketService {
  private socket: Socket | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;
  private listeners: Map<string, Set<(data: any) => void>> = new Map();

  connect(token?: string) {
    if (this.socket?.connected) {
      return;
    }

    const url = token ? `${WS_URL}?token=${token}` : WS_URL;

    this.socket = io(url, {
      transports: ['websocket'],
      reconnection: true,
      reconnectionDelay: this.reconnectDelay,
      reconnectionAttempts: this.maxReconnectAttempts,
    });

    this.socket.on('connect', () => {
      console.log('WebSocket connected');
      this.reconnectAttempts = 0;
      this.emit('connected', {});
    });

    this.socket.on('disconnect', () => {
      console.log('WebSocket disconnected');
    });

    this.socket.on('error', (error) => {
      console.error('WebSocket error:', error);
    });

    this.socket.on('reconnect', (attemptNumber) => {
      console.log('WebSocket reconnected after', attemptNumber, 'attempts');
      this.reconnectAttempts = 0;
    });

    this.socket.on('reconnect_attempt', (attemptNumber) => {
      console.log('WebSocket reconnect attempt', attemptNumber);
      this.reconnectAttempts = attemptNumber;
    });

    this.socket.on('reconnect_failed', () => {
      console.error('WebSocket reconnection failed');
    });

    // Set up event listeners
    this.setupEventListeners();
  }

  private setupEventListeners() {
    if (!this.socket) return;

    // Listen for all events and forward to registered listeners
    const events = [
      'call.started',
      'call.updated',
      'call.ended',
      'interaction.added',
      'qa.score.updated',
      'sentiment.changed',
      'escalation.triggered',
      'escalation.completed',
      'notification.created',
    ];

    events.forEach((event) => {
      this.socket?.on(event, (data) => {
        const listeners = this.listeners.get(event);
        if (listeners) {
          listeners.forEach((listener) => listener(data));
        }
      });
    });
  }

  subscribe(event: string, callback: (data: any) => void) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }
    this.listeners.get(event)?.add(callback);

    // If subscribing to a specific call, emit subscribe event
    if (event.startsWith('call.') && this.socket?.connected) {
      this.emit('subscribe', { type: 'call', event });
    }
  }

  unsubscribe(event: string, callback: (data: any) => void) {
    const listeners = this.listeners.get(event);
    if (listeners) {
      listeners.delete(callback);
      if (listeners.size === 0) {
        this.listeners.delete(event);
      }
    }
  }

  subscribeToCall(callId: string) {
    if (this.socket?.connected) {
      this.emit('subscribe', { type: 'call', call_id: callId });
    }
  }

  unsubscribeFromCall(callId: string) {
    if (this.socket?.connected) {
      this.emit('unsubscribe', { type: 'call', call_id: callId });
    }
  }

  subscribeToAllCalls() {
    if (this.socket?.connected) {
      this.emit('subscribe', { type: 'all_calls' });
    }
  }

  emit(event: string, data: any) {
    if (this.socket?.connected) {
      this.socket.emit(event, data);
    }
  }

  disconnect() {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
      this.listeners.clear();
    }
  }

  isConnected(): boolean {
    return this.socket?.connected || false;
  }
}

export const websocketService = new WebSocketService();

File: src/api/routes/websocket.py
"""WebSocket routes for real-time updates"""

import socketio
from typing import Optional
import jwt
from jose import JWTError
from src.utils.config import get_settings
from src.utils.logging import get_logger

logger = get_logger(__name__)
settings = get_settings()

# Create Socket.IO server
sio = socketio.AsyncServer(
    cors_allowed_origins="*",
    async_mode='asgi'
)

# Store connected clients and their user IDs
connected_clients = set()
client_user_map = {}  # Map sid -> user_id


def get_token_from_query(query_string: str) -> Optional[str]:
    """Extract token from query string"""
    if not query_string:
        return None
    params = dict(param.split('=') for param in query_string.split('&') if '=' in param)
    return params.get('token')


def verify_token(token: str) -> bool:
    """Verify JWT token"""
    try:
        jwt.decode(token, settings.JWT_SECRET_KEY, algorithms=[settings.JWT_ALGORITHM])
        return True
    except JWTError:
        return False


def get_user_id_from_token(token: str) -> Optional[str]:
    """Extract user ID from JWT token"""
    try:
        payload = jwt.decode(token, settings.JWT_SECRET_KEY, algorithms=[settings.JWT_ALGORITHM])
        return payload.get("sub")
    except JWTError:
        return None


@sio.event
async def connect(sid, environ, auth):
    """Handle client connection"""
    query_string = environ.get('QUERY_STRING', '')
    token = get_token_from_query(query_string)
    
    # Enforce authentication based on environment
    if settings.APP_ENV == "production":
        # In production, require authentication
        if not token:
            logger.warning("websocket_auth_missing", sid=sid)
            return False
        if not verify_token(token):
            logger.warning("websocket_auth_failed", sid=sid)
            return False
    else:
        # In development, allow connections without token but verify if provided
        if token and not verify_token(token):
            logger.warning("websocket_auth_failed", sid=sid)
            return False
    
    # Extract user ID from token and store mapping
    if token:
        user_id = get_user_id_from_token(token)
        if user_id:
            client_user_map[sid] = user_id
            # Automatically subscribe to user's notification room
            await sio.enter_room(sid, f"user:{user_id}")
            logger.info("websocket_user_subscribed", sid=sid, user_id=user_id)
    
    connected_clients.add(sid)
    logger.info("websocket_client_connected", sid=sid)
    return True


@sio.event
async def disconnect(sid):
    """Handle client disconnection"""
    connected_clients.discard(sid)
    # Clean up user mapping
    if sid in client_user_map:
        user_id = client_user_map.pop(sid)
        await sio.leave_room(sid, f"user:{user_id}")
    logger.info("websocket_client_disconnected", sid=sid)


@sio.event
async def subscribe(sid, data):
    """Handle subscription requests"""
    subscription_type = data.get('type')
    
    if subscription_type == 'all_calls':
        await sio.enter_room(sid, 'all_calls')
        logger.info("websocket_subscribed", sid=sid, room='all_calls')
    elif subscription_type == 'call' and data.get('call_id'):
        room = f"call:{data['call_id']}"
        await sio.enter_room(sid, room)
        logger.info("websocket_subscribed", sid=sid, room=room)
    elif subscription_type == 'notifications':
        # Subscribe to notifications for the authenticated user
        # User ID should be extracted from token during connection
        # For now, we'll use a general notifications room
        await sio.enter_room(sid, 'notifications')
        logger.info("websocket_subscribed", sid=sid, room='notifications')


@sio.event
async def unsubscribe(sid, data):
    """Handle unsubscription requests"""
    subscription_type = data.get('type')
    
    if subscription_type == 'all_calls':
        await sio.leave_room(sid, 'all_calls')
    elif subscription_type == 'call' and data.get('call_id'):
        room = f"call:{data['call_id']}"
        await sio.leave_room(sid, room)
    
    logger.info("websocket_unsubscribed", sid=sid)


# Event emitter functions
async def emit_call_started(call_data: dict):
    """Emit call.started event"""
    await sio.emit('call.started', {'call': call_data}, room='all_calls')
    await sio.emit('call.started', {'call': call_data}, room=f"call:{call_data.get('id')}")


async def emit_call_updated(call_data: dict):
    """Emit call.updated event"""
    await sio.emit('call.updated', {'call': call_data}, room='all_calls')
    await sio.emit('call.updated', {'call': call_data}, room=f"call:{call_data.get('id')}")


async def emit_call_ended(call_id: str):
    """Emit call.ended event"""
    await sio.emit('call.ended', {'call_id': call_id}, room='all_calls')
    await sio.emit('call.ended', {'call_id': call_id}, room=f"call:{call_id}")


async def emit_interaction_added(call_id: str, interaction_data: dict):
    """Emit interaction.added event"""
    await sio.emit('interaction.added', {
        'call_id': call_id,
        'interaction': interaction_data
    }, room='all_calls')
    await sio.emit('interaction.added', {
        'call_id': call_id,
        'interaction': interaction_data
    }, room=f"call:{call_id}")


async def emit_qa_score_updated(call_id: str, qa_score_data: dict):
    """Emit qa.score.updated event"""
    await sio.emit('qa.score.updated', {
        'call_id': call_id,
        'qa_score': qa_score_data
    }, room='all_calls')
    await sio.emit('qa.score.updated', {
        'call_id': call_id,
        'qa_score': qa_score_data
    }, room=f"call:{call_id}")


async def emit_sentiment_changed(call_id: str, sentiment_data: dict):
    """Emit sentiment.changed event"""
    await sio.emit('sentiment.changed', {
        'call_id': call_id,
        'sentiment': sentiment_data
    }, room='all_calls')
    await sio.emit('sentiment.changed', {
        'call_id': call_id,
        'sentiment': sentiment_data
    }, room=f"call:{call_id}")


async def emit_escalation_triggered(call_id: str, escalation_data: dict):
    """Emit escalation.triggered event"""
    await sio.emit('escalation.triggered', {
        'call_id': call_id,
        'escalation': escalation_data
    }, room='all_calls')
    await sio.emit('escalation.triggered', {
        'call_id': call_id,
        'escalation': escalation_data
    }, room=f"call:{call_id}")


async def emit_escalation_completed(call_id: str, escalation_data: dict):
    """Emit escalation.completed event"""
    await sio.emit('escalation.completed', {
        'call_id': call_id,
        'escalation': escalation_data
    }, room='all_calls')
    await sio.emit('escalation.completed', {
        'call_id': call_id,
        'escalation': escalation_data
    }, room=f"call:{call_id}")


async def emit_notification_created(user_id: str, notification_data: dict):
    """Emit notification.created event to specific user"""
    # Emit to user-specific room
    await sio.emit('notification.created', {
        'notification': notification_data
    }, room=f"user:{user_id}")
    # Also emit to general notifications room for any listeners
    await sio.emit('notification.created', {
        'notification': notification_data
    }, room='notifications')

